Add login link template override for unverified & verified verbiage

B77Mills opened the pull request:
Add the ability to override the verbiage that appears above the login link on a per app or appContext basis.  This will, have the ability to support simple html element like `<b>`, `<strong>`, `<em>`, `<i>`, `<u>`

For now the ability to subject line is there, but hidden behind a display: none class on the form row above the unverifed field

<img width="917" alt="Screen Shot 2023-07-11 at 1 22 08 PM" src="https://github.com/parameter1/identity-x/assets/3845869/50f6680c-1c64-4772-b2da-40f6f33d68ba">

<img width="917" alt="Screen Shot 2023-07-11 at 1 22 21 PM" src="https://github.com/parameter1/identity-x/assets/3845869/3a2b38d2-d26e-40fb-9f1e-e388397bea62">

### unverified
<img width="846" alt="Screen Shot 2023-07-11 at 1 40 26 PM" src="https://github.com/parameter1/identity-x/assets/3845869/7e8a9e68-e44e-4194-ad6c-8ecfb7257658">

### verified
<img width="704" alt="Screen Shot 2023-07-11 at 1 27 16 PM" src="https://github.com/parameter1/identity-x/assets/3845869/bc6d7b87-0edf-4bb3-840f-29491e95e63c">

B77Mills committed:
Add striptags package

B77Mills committed:
Add loginLinkTemplate schema, definitions and resolvers to graphql

B77Mills committed:
Add loginLinkTemplate inputs to manage app/appContext view

Add handeling for update/add create of

B77Mills committed:
Ensure the required language prop is on the fragment

B77Mills committed:
Update SendLoginLink Action & templates

Update sendLoginLink action to account for new loginLinkTemplate overrides.

B77Mills committed:
Update how loginLinkTemplate is set.

B77Mills committed:
fix spreadding of loginLinkTemplate objects

B77Mills committed:
Convert subjectLine prop to subject

B77Mills committed:
Update login email inputs labels & hide subject for now

B77Mills committed:
udpate template

B77Mills committed:
Missed this switch from subjectLine to subject

B77Mills committed:
update list of supported html elments

zarathustra323 commented on services/application/src/email-templates/login-link/en-us.js:
```diff
@@ -1,45 +1,56 @@
+  user,
+} = {}) => {
+  const { verified } = user;
+  const subject = loginLinkTemplate.subject ? loginLinkTemplate.subject : 'Your personal login link';
```
`const subject = loginLinkTemplate.subject || 'Your personal login link';`

zarathustra323 commented on services/application/src/email-templates/login-link/en-us.js:
```diff
@@ -1,45 +1,56 @@
+} = {}) => {
+  const { verified } = user;
+  const subject = loginLinkTemplate.subject ? loginLinkTemplate.subject : 'Your personal login link';
+  const unverifiedVerbiage = loginLinkTemplate.unverifiedVerbiage ? loginLinkTemplate.unverifiedVerbiage : `You recently requested to log in to <strong>${appName}</strong>. This link is good for one hour.`;
```
Use the `||` syntax here as well

zarathustra323 commented on services/application/src/email-templates/login-link/en-us.js:
```diff
@@ -1,45 +1,56 @@
+  const { verified } = user;
+  const subject = loginLinkTemplate.subject ? loginLinkTemplate.subject : 'Your personal login link';
+  const unverifiedVerbiage = loginLinkTemplate.unverifiedVerbiage ? loginLinkTemplate.unverifiedVerbiage : `You recently requested to log in to <strong>${appName}</strong>. This link is good for one hour.`;
+  const verifiedVerbiage = loginLinkTemplate.verifiedVerbiage ? loginLinkTemplate.verifiedVerbiage : `You recently requested to log in to <strong>${appName}</strong>. This link is good for one hour.`;
```
And here too :)

zarathustra323 commented on services/application/src/actions/user/send-login-link.js:
```diff
@@ -56,24 +56,34 @@ module.exports = async ({
 
+  const loginLinkTemplate = {
+    ...app.loginLinkTemplate.toObject(),
+    ...context.loginLinkTemplate.toObject(),
```
Based on L52 above, it's possible the the `context` object, in some cases, will be a POJO, not a Mongoose object. As such, the `toObject` call will throw an error... a somewhat "hacky fix" would be to add a `toObject` function to the empty fallback on L52 that returns an empty object.... `{ toObject: ()  => ({}) };` this is probably the quickest way to deal with it

zarathustra323 commented on services/application/src/email-templates/login-link/en-us.js:
```diff
@@ -1,45 +1,56 @@
 
-Login to ${appName} by visiting this link:
-${url}
+  Login to ${appName} by visiting this link:
```
All the added spaces to the text email should be removed. whitespace is preserved in text emails

zarathustra323 commented on services/application/src/email-templates/login-link/es-mx.js:
```diff
@@ -1,34 +1,44 @@
+  user,
+} = {}) => {
+  const { verified } = user;
+  const subject = loginLinkTemplate.subject ? loginLinkTemplate.subject : 'Su enlace de inicio de sesión personal';
```
Do the same `||` logic as in the english version

zarathustra323 commented on services/application/src/mongodb/schema/application.js:
```diff
@@ -1,8 +1,38 @@
+const loginLinkTemplate = new Schema({
+  subject: {
+    type: String,
+    set: (v) => {
```
The three setters can change to `set: stripHtml,` -- since `stripHtml` is doing the exact same logic

zarathustra323 commented on services/manage/app/mixins/action-mixin.js:
```diff
@@ -26,6 +26,14 @@ export default Mixin.create(LoadingMixin, {
     this.hideLoading();
   },
 
+  validateLoginLinkTemplateObj(loginLinkTemplate = {}){
```
Is this necessary?

solocommand replied:
Agreed, this should be removed. There shouldn't be a need to validate and modify the value here, so long as it is returned from the GraphQL query properly, and initialized on create so that the inputs can bind properly. Since this is being used on submission, that would seem to have already been accounted for.

zarathustra323 replied:
```js
formatLoginLinkTemplateInput({ subject, unverifiedVerbiage, verifiedVerbiage } = {}){
  return {
    subject,
    unverifiedVerbiage,
    verifiedVerbiage,
  };
},
```

B77Mills committed:
Add toObject function to fallback context object

This will ensure they fallback object has the toObject function like the mongoose object does

B77Mills committed:
Clean up email templating  logic

B77Mills committed:
Update set to be stripHtml

this is already doing a null check

solocommand commented on services/application/src/email-templates/login-link/en-us.js:
```diff
@@ -1,34 +1,44 @@
+      </head>
+      <body>
+        <p>${verbiage}</p>
+        <p><a href="${url}">Login to ${appName}</a></p>
```
`Login` should be `Log in` here and on L43

solocommand commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Can we have the wording changed on the login email?](https://trello.com/c/roxLrXd6/381-can-we-have-the-wording-changed-on-the-login-email)

B77Mills committed:
Update to use formatLoginLinkTemplateInput

Update the function being used to format the loginLinkTemplateInput when modifiying appContexts

B77Mills committed:
Switch `Login to`  => `Log in to`

B77Mills committed:
Handle setting of fallback loginLinkTemplate values

 try context, then app per key

B77Mills committed:
trim loginLinkTemplate values
